Fixes #19067 - React DiffView Component

glekner opened the pull request:
Creating a React component for viewing Diff (right now in Audits page)
Progress so far:


![diff](https://user-images.githubusercontent.com/24938324/46917448-22df3200-cfd0-11e8-9733-fd2d1e9fca1b.gif)



Need suggestions for styling, maximum height before scrolling etc.
Using the react component `react-diff-view` which promises syntax highlighting in v2.0, not sure when will it come out but it can be a nice feature in the future.
This pr can help us migrate the whole Audits page to React.
<!---

Thank you for contributing to The Foreman, please read the
[following guide](https://www.theforeman.org/contribute.html), in short:

* [Create an issue](https://projects.theforeman.org/projects/foreman/issues)
* Reference the issue via `Fixes #1234` in the commit message
* Prefer present-tense, imperative-style commit messages
* Mark all strings for translation, see [Translating section in the guide]
(https://projects.theforeman.org/projects/foreman/wiki/Translating)
* Prepend `[WIP]` for work in progress to prevent bots from triggering actions
* Be patient, we will do our best to take a look as soon as we can
* Explain the purpose of the PR, attach screenshots if applicable
* Be nice and respectful

We are running bots that will poke you if you miss an item from the list :-)

--->

glekner committed:
Fixes #19067 - React DiffView Component

terezanovotna commented:
This looks great @glekner!

I think it's a good idea to set max heigh before scrolling. How about 400px? 
Could the buttons (Split, Unified) have the same width?
I like how you created it almost identical as it is in GitHub, great job!

glekner commented:
@terezanovotna 
- max-height is now 400px
- buttons width set to 60px

glekner commented:
- Moved Radio buttons to the container so they won't disappear upon scrolling
- Added more styling + font rules to match Github even more 💯 
<img width="994" alt="screenshot 2018-10-15 16 03 52" src="https://user-images.githubusercontent.com/24938324/46952694-4370c000-d094-11e8-8fa5-4ffa669c13ff.png">

<img width="934" alt="screenshot 2018-10-15 16 08 32" src="https://user-images.githubusercontent.com/24938324/46952787-977ba480-d094-11e8-943e-110361e62ac1.png">

terezanovotna commented:
Looks great! Interested in seeing the storybook when you have it!

glekner commented:
Thanks @terezanovotna 
Right now deploying the storybook to github deploys the old storybook for some reason.
I did add a story though, so you can clone my branch and run storybook locally to see it.

ohadlevy commented on webpack/assets/javascripts/react_app/components/DiffView/DiffContainer.js:
```diff
@@ -0,0 +1,61 @@
+  constructor(props) {
+    super(props);
+    this.state = {
+      viewType: 'split',
```
the default is defined twice, here and in the DiffView component, should we define it only once (perhaps here?)

ohadlevy commented:
@kgaikwad can you have a look? thanks!

amirfefer commented on webpack/assets/javascripts/react_app/components/AuditsList/ExpansiveView.js:
```diff
@@ -1,5 +1,6 @@
 import React from 'react';
 import PropTypes from 'prop-types';
+import DiffContainer from '../DiffView/DiffContainer';
 import { renderTemplatesDiff } from '../../../foreman_editor';
```
isn't it redundant now?
also the `componentDidMount` callback with `renderTemplatesDiff` and the ref.

amirfefer commented on webpack/assets/javascripts/react_app/components/DiffView/DiffContainer.js:
```diff
@@ -0,0 +1,61 @@
+  }
+
+  onUnifiedClick() {
+    this.setState({ viewType: 'unified' });
```
wouldn't it be better as a redux store prop?
with it we can handle states outside of the component's scope.

amirfefer commented on webpack/assets/javascripts/react_app/components/DiffView/DiffContainer.js:
```diff
@@ -0,0 +1,61 @@
+  }
+
+  onSplitClick() {
+    this.setState({ viewType: 'split' });
```
ditto

amirfefer commented on webpack/assets/javascripts/react_app/components/DiffView/DiffContainer.js:
```diff
@@ -0,0 +1,61 @@
+        <div id="diff-radio-buttons">
+          <ButtonGroup>
+            <Button
+              className={this.state.viewType === 'split' ? 'diff-button active' : 'diff-button'}
```
you can use `classnames` for a conditional class
https://github.com/JedWatson/classnames

amirfefer reviewed (changes requested):
Thanks @glekner !
Looks great :+1: 
I left few inline comments

glekner commented:
- applied @amirfefer suggestions. still keeping redux out.
- included `unidiff` in webpack to fix tests
- reverted some of #5869 editor changes, a future react-ace PR should remove this file hopefully.

ohadlevy commented:
@glekner looks like https://ci.theforeman.org/job/test_develop_pr_core/19943/database=postgresql,ruby=2.3,slave=fast/testReport/junit/(root)/AuditMailerTest/test_0006_Audit_template_change_should_not_crash/ is test failure.

kgaikwad commented on webpack/assets/javascripts/react_app/components/AuditsList/ExpansiveView.js:
```diff
@@ -1,6 +1,6 @@
 import PropTypes from 'prop-types';
-import { renderTemplatesDiff } from '../../../foreman_editor';
+import DiffContainer from '../DiffView/DiffContainer';
 import { translate as __ } from '../../common/I18n';
```
we can remove this line as  I think we don't require it in this file.

kgaikwad replied:
@ohadlevy,
I can see that translation used for actions like Diff, Preview on template page.
Do we need to apply translate about 'split', 'unified'?
I think yes.

glekner replied:
@kgaikwad we use it to translate "Comments" line 78.
good point on the split, unified. ill translate them

kgaikwad reviewed (changes requested):
@glekner,
+1, looks great!

Added just small inline comment.

glekner commented:
@ohadlevy `audit_mailer_test.rb` is passing locally.

ohadlevy commented on package.json:
```diff
@@ -88,6 +88,7 @@
     "react": "^16.4.0",
     "react-bootstrap": "0.32.1",
     "react-debounce-input": "^3.2.0",
+    "react-diff-view": "^1.8.1",
```
packaging PR at https://github.com/theforeman/foreman-packaging/pull/3136

ohadlevy commented on package.json:
```diff
@@ -102,6 +103,7 @@
     "reselect": "^3.0.1",
     "seamless-immutable": "^7.1.2",
     "select2": "~3.5.2-browserify",
+    "unidiff": "0.0.4",
```
https://github.com/theforeman/foreman-packaging/pull/3137

ohadlevy commented on webpack/assets/javascripts/react_app/components/DiffView/DiffRadioButtons.js:
```diff
@@ -0,0 +1,35 @@
+  <div id="diff-radio-buttons">
+    <ButtonGroup>
+      {radioButton(stateView, 'split', changeState)}
+      {radioButton(stateView, 'unified', changeState)}
```
I think you should define the translation here, e.g.``` __('split')``` ,this is important for two reasons:
1.```translate``` is not a keyword that our string extraction code catches /cc @amirfefer IMHO docs needs to reflect this.
2. this approach will not actually extract the string, just translate it, hence it will never actually show it as translated.

ohadlevy commented on webpack/assets/javascripts/react_app/components/DiffView/DiffView.stories.js:
```diff
@@ -0,0 +1,15 @@
+import React from 'react';
```
did you publish the storybook somewhere?

glekner replied:
[Storybook](https://glekner.github.io/foreman/?knob-Old%20Text=Old%20Text&knob-New%20Text=New%20Text&selectedKind=Components%2FDiffView&selectedStory=DiffView&full=0&addons=1&stories=1&panelRight=0&addonPanel=storybooks%2Fstorybook-addon-knobs) @terezanovotna

amirfefer commented on webpack/assets/javascripts/react_app/components/DiffView/DiffView.js:
```diff
@@ -0,0 +1,37 @@
+const getDiff = (oldText, newText) => {
+  const diffText = formatLines(diffLines(oldText, newText), { context: 3 });
+  // these two lines are faked to mock git diff output
+  const header = ['diff --git a/a b/b', 'index 0000000..1111111 100644'];
```
what about using a dedicated tool for creating git-diff, such as https://github.com/danday74/git-diff ?
`unidiff` looks a bit fragile

ohadlevy commented:
note that we also currently use js-diff in the original/current editor diff.

On Sun, Oct 28, 2018 at 3:21 PM Amir Fefer <notifications@github.com> wrote:

> *@amirfefer* commented on this pull request.
> ------------------------------
>
> In webpack/assets/javascripts/react_app/components/DiffView/DiffView.js
> <https://github.com/theforeman/foreman/pull/6138#discussion_r228748335>:
>
> > @@ -0,0 +1,37 @@
> +import React from 'react';
> +import PropTypes from 'prop-types';
> +
> +import { parseDiff, Diff, markCharacterEdits } from 'react-diff-view';
> +import { formatLines, diffLines } from 'unidiff';
> +import 'react-diff-view/index.css';
> +import './diffview.scss';
> +
> +const getDiff = (oldText, newText) => {
> +  const diffText = formatLines(diffLines(oldText, newText), { context: 3 });
> +  // these two lines are faked to mock git diff output
> +  const header = ['diff --git a/a b/b', 'index 0000000..1111111 100644'];
>
> what about using a dedicated tool for creating git-diff, such as
> https://github.com/danday74/git-diff ?
> unidiff looks a bit fragile
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/theforeman/foreman/pull/6138#pullrequestreview-169092172>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AABOx_8D-e17JkqjwMprTtFGTD7pTSb_ks5upa9qgaJpZM4XbInE>
> .
>

glekner commented:
@ohadlevy @amirfefer tried `git-diff` but it crashes on compilation.

@otakustay and me discussed this and he tried to help with a standalone file
https://pastebin.com/ibBq39Yt
we can import `formatLines, diffLines` from this file.
we can call it `foreman_diff.js` or something similar, its 30kb and replaces `unidiff`.

i'm working meanwhile on a redux editor which doesn't need `jsdiff`  anylonger. i'm using this pr in the new editor so i'm waiting for this pr to get merged. thoughts?

ohadlevy commented:
> @otakustay and me discussed this and he tried to help with a standalone file
> https://pastebin.com/ibBq39Yt
> we can import `formatLines, diffLines` from this file.
> we can call it `foreman_diff.js` or something similar, its 30kb and replaces `unidiff`.

it would be best if you guys can package it as a npm package, this way we won't need to carry it just for foreman, and perhaps other people will find it useful too? :)

ohadlevy commented on config/webpack.config.js:
```diff
@@ -80,8 +80,8 @@ module.exports = env => {
-          // Include novnc in webpack, transpiling is needed for phantomjs (which does not support ES6) to run tests
-          exclude: /node_modules(?!\/@novnc)/,
+          // Include novnc, unidiff in webpack, transpiling is needed for phantomjs (which does not support ES6) to run tests
+          exclude: /node_modules(?!\/(@novnc|unidiff))/,
```
maybe worth adding a link as TODO to the GH issue you mentioned about updating unidiff that it doesnt require this workaround?

ohadlevy commented:
@ekohl ack?

glekner commented:
replaced `unidiff` with a fork `unidiff-bundle` for browser support

ohadlevy commented:
Don't you need to remove the webpack workaround? I'll create a new
packaging PR if you can confirm

On Thu, Nov 1, 2018, 12:02 PM Gilad Lekner <notifications@github.com wrote:

> replaced unidiff with a fork unidiff-bundle for browser support
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/theforeman/foreman/pull/6138#issuecomment-434992260>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AABOx82o7Viiq8ZibkyvcRTONQJpBOPvks5uqsa_gaJpZM4XbInE>
> .
>

glekner commented:
TFM is still undefined.
@otakustay said he bundled into UMD so it should work without webpack but
the tests still fail, so I kept the rule.

ohadlevy commented:
So I'm not sure what's the value then?

On Fri, Nov 2, 2018, 10:06 AM Gilad Lekner <notifications@github.com wrote:

> TFM is still undefined.
> @otakustay said he bundled into UMD so it should work without webpack but
> the tests still fail, so I kept the rule.
>
> On Fri, 2 Nov 2018 at 9:57 Ohad Levy <notifications@github.com> wrote:
>
> > Don't you need to remove the webpack workaround? I'll create a new
> > packaging PR if you can confirm
> >
> > On Thu, Nov 1, 2018, 12:02 PM Gilad Lekner <notifications@github.com
> > wrote:
> >
> > > replaced unidiff with a fork unidiff-bundle for browser support
> > >
> > > —
> > > You are receiving this because you modified the open/close state.
> > > Reply to this email directly, view it on GitHub
> > > <
> https://github.com/theforeman/foreman/pull/6138#issuecomment-434992260
> > >,
> > > or mute the thread
> > > <
> >
> https://github.com/notifications/unsubscribe-auth/AABOx82o7Viiq8ZibkyvcRTONQJpBOPvks5uqsa_gaJpZM4XbInE
> > >
> > > .
> > >
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/theforeman/foreman/pull/6138#issuecomment-435299840
> >,
> > or mute the thread
> > <
> https://github.com/notifications/unsubscribe-auth/AXyHVBR3wEvF1EwyNNngn_Rl_1xVNyqvks5uq_r5gaJpZM4XbInE
> >
> > .
> >
> --
> *בברכה, *
>
> *גלעד לקנר*
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/theforeman/foreman/pull/6138#issuecomment-435301404>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AABOx5c21xEBXE77QrAhfv5pa2vJ4uooks5uq_zrgaJpZM4XbInE>
> .
>

glekner commented:
There is no value right now. I thought we should still use the fork because he is more responsive and wants to help. Can we merge and remove unidiff from webpack in a future PR?

glekner commented:
@ohadlevy rebased and changed back to `unidiff`

ohadlevy commented:
@glekner I would assume that in order to make that work @otakustay would also need babel support...?

ohadlevy commented:
@glekner as it stands now, the package is not usable, as it has incorrect name in package.json, making our build system to fail, I would suggest to keep the current one until ```undiff-bundle``` is ready/usable?

ohadlevy commented:
test failures not related.

ekohl reviewed (approved):
:+1: on the packaging side of it.
